### PR TITLE
Add a tool to convert other recipes to mbd recipes and dynamic recipe

### PR DESCRIPTION
**Pull request by vfyjxf**
Added a tool that allows script writers to convert recipes from other RecipeType to mbd recipes.

Below are two example scripts which convert smelting recipes and crafting recipes into mbd recipes.
``` js
var Ingredient = java('net.minecraft.world.item.crafting.Ingredient')

onEvent('mbd.recipe_converter_register', event => {
    console.info("Starting conversion of recipes")
    event.register('minecraft:smelting', 'converter_map', (recipe, map) => {
        map.start().duration(1)
            .inputItems(ArrayUtils.create(recipe.getIngredients(), Ingredient))
            .outputItems(recipe.getResultItem())
            .buildAndRegister();
    })
    event.register('minecraft:crafting', 'crafting_map', (recipe, map) => {
        map.start().duration(1)
            .inputItems(ArrayUtils.create(recipe.getIngredients(), Ingredient))
            .outputItems(recipe.getResultItem())
            .buildAndRegister();
    })
    console.info("Finished conversion of recipes")
})
```

### CODE REVIEW

**Comment by vfyjxf** [comment]
I accidentally submitted ArrayUtils directly into LDLib because I had set the remote address incorrectly. This tool class helps script writers convert Lists to arrays.

**Comment by vfyjxf** [comment]
now modpacker can use SetupRecipeEvent and SearchRecipeEvent to create dynamic recipe to their machine.
Here is a example:
It will remove chance input when multiblock has netherite_block, has copper_block will make input increase,
glod block will cost two iron ingot to decrease the time.
```js
onEvent('mbd.setup_recipe.tester.updatable', event => {
    if (event.getRecipe() == null || event.getRecipe().dynamic) return
    console.info("Ok,this a updatable multiblock machine. I'm going to change the recipe")
    let componentData = event.getComponentData()
    let copperAmount = componentData.get(Block.getBlock('minecraft:copper_block'))
    let ironAmount = componentData.get(Block.getBlock('minecraft:iron_block'))
    let goldAmount = componentData.get(Block.getBlock('minecraft:gold_block'))
    let diamondAmount = componentData.get(Block.getBlock('minecraft:diamond_block'))
    let netheriteAmount = componentData.get(Block.getBlock('minecraft:netherite_block'))
    let handler = event.getHandlerFromRecipe()
    let duration = diamondAmount > 0 ? event.getRecipe().duration / (diamondAmount * 2) : event.getRecipe().duration;
    if (netheriteAmount > 0) {
        handler.removeTypedInputs(false, MbdRegistry.getCapability('item'), content => {
            return content.chance < 1
        })
        handler.outputHandler(map => {
            let contents = map.get('item').getRight()
            for (let i = 0; i < contents.size(); i++) {
                var content = contents.get(i)
                var nextChance = content.chance * 2
                nextChance = nextChance > 1 ? 1 : nextChance
                content.chance = nextChance
            }
        })
    }
    if (copperAmount > 0) {
        console.info("InputAdder Test")
        handler.inputAdder(false, 'item', copperAmount * 2)
    }
    if (goldAmount > 0) {
        let iron = Item.of('minecraft:iron_ingot', 2)
        handler.inputItems(iron)
        duration = 5
    }
    handler.duration(duration)
    event.setRecipe(handler.apply())

})
```

File: src/main/java/com/lowdragmc/multiblocked/api/pattern/BlockPattern.java [hunk 1]
```diff
@@ -152,6 +147,11 @@ public boolean checkPatternAt(MultiblockState worldState, BlockPos centerPos, Di
                                 z++;//continue searching for the first aisle
                             }
                             continue loop;
+                        } else {
+                            //components statistics
+                            worldState.getMatchContext()
+                                    .getOrCreate("components", HashMap<Block, Integer>::new)
```
Comment (Yefancy) [inline src/main/java/com/lowdragmc/multiblocked/api/pattern/BlockPattern.java]: Block->BlockState

**Review by Yefancy — CHANGES_REQUESTED** [review]

**Comment by vfyjxf** [comment]
A new recipe condition has been added. modpacker can use it to create some strange requirement,such a machine require a pig near it.
Here is a example to create a miner:
```js
onEvent('recipes', event => {
    event.recipes.multiblocked.multiblock("miner")
        .setChance(0.5)
        .outputItem('2x minecraft:netherite_ingot')
        .predicate((recipe, recipeLogic) => {
            let under = getUnder(recipeLogic)
            let block = Block.getBlock('minecraft:netherite_block')
            let b = block.equals(under)
            return b
        },Component.string('Require a netherite block under the controller'))
        .duration(60)

    event.recipes.multiblocked.multiblock("miner")
        .setChance(0.5)
        .outputItem('2x minecraft:diamond')
        .predicate((recipe, recipeLogic) => {
            let under = getUnder(recipeLogic)
            let b = Block.getBlock('minecraft:diamond_block').equals(under)
            return b
        },Component.string('Require a diamond block under the controller'))
        .duration(60)

    event.recipes.multiblocked.multiblock("miner")
        .setChance(0.5)
        .outputItem('2x minecraft:iron_ingot')
        .predicate((recipe, recipeLogic) => {
            let under = getUnder(recipeLogic)
            let b = Block.getBlock('minecraft:iron_block').equals(under)
            return b
        },Component.string('Require a iron block under the controller'))
        .duration(60)

    event.recipes.multiblocked.multiblock("miner")
        .setChance(0.5)
        .outputItem('2x minecraft:gold_ingot')
        .predicate((recipe, recipeLogic) => {
            let under = getUnder(recipeLogic)
            let block = Block.getBlock('minecraft:gold_block')
            let b = block.equals(under)
            return b
        },Component.string('Require a gold block under the controller'))
        .duration(60)
})

function getUnder(recipeLogic) {
    let controllerTe = recipeLogic.controller.self();
    let level = controllerTe.getLevel()
    return level.getBlockState(controllerTe.getBlockPos().below()).getBlock();
}

```

**Comment by vfyjxf** [comment]
Incidentally, I didn't add predicate conditional textures because I don't have the time and I'm not good at drawing textures.

**Comment by vfyjxf** [comment]
To add to this, I have also made some extensions to the RecipeFinishEvent, which now has two child events, Pre and Post, using the Post event to safely modify the structure of multiblocks after the recipe has been completed.
Example:
```js

onEvent("mbd.recipe_finish.post.tester.updatable", event => {
    let gold = Block.getBlock('minecraft:gold_block')
    let diamond = Block.getBlock('minecraft:diamond_block')
    let netherite = Block.getBlock('minecraft:netherite_block')
    let controller = event.getRecipeLogic().controller
    let iron = Block.getBlock('minecraft:iron_block')
    let level = event.getRecipeLogic().controller.self().getLevel()

    let poses = controller.getMultiblockState().getCache()
    for (const pos of poses) {
        var find = false
        var block = level.getBlockState(pos).getBlock()
        if (diamond.equals(block)) {
            level.setBlock(pos, gold.defaultBlockState(), 3)
            break;
        }
        if (netherite.equals(block)) {
            level.setBlock(pos, diamond.defaultBlockState(), 3)
            break;
        }

    }
})

```

**Review by Yefancy — APPROVED** [review]

**Pull request merged**